src/linear_algebra/special_linear_group.lean

```diff
@@ -0,0 +1,145 @@
+  Author: Anne Baanen.
+
+  The Special Linear group $SL(n, R)$
+-/
+import linear_algebra.basic
+import linear_algebra.matrix
+import linear_algebra.nonsingular_inverse
+import tactic.norm_cast
+
+/-!
+# The Special Linear group $SL(n, R)$
+
+This file defines the elements of the Special Linear group `special_linear_group n R`,
+also written `SL(n, R)` or `SLₙ(R)`, consisting of all `n` by `n` `R`-matrices with
+determinant `1`.  In addition, we define the group structure on `special_linear_group n R`
+and the embedding into the general linear group `general_linear_group R (n → R)`
+(i.e. `GL(n, R)` or `GLₙ(R)`).
+
+## Main definitions
+
+ * `matrix.special_linear_group` is the type of matrices with determinant 1
+ * `matrix.special_linear_group.group` gives the group structure (under multiplication)
+ * `matrix.special_linear_group.embedding_GL` is the embedding `SLₙ(R) → GLₙ(R)`
+
+## Implementation notes
+The inverse operation in the `special_linear_group` is defined to be the adjugate
+matrix, so that `special_linear_group n R` has a group structure for all `comm_ring R`.
+
+We define the elements of `special_linear_group` to be matrices, since we need to
+compute their determinant. This is in contrast with `general_linear_group R M`,
+which consists of invertible `R`-linear maps on `M`.
+
+## References
+
+ * https://en.wikipedia.org/wiki/Special_linear_group
+
+## Tags
+
+matrix group, group, matrix inverse
+-/
+
+namespace matrix
+universes u v
+open_locale matrix
+open linear_map
+
+set_option class.instance_max_depth 60
+
+section
+
+variables (n : Type u) [fintype n] [decidable_eq n] (R : Type v) [comm_ring R]
+
+/-- `special_linear_group n R` is the group of `n` by `n` `R`-matrices with determinant equal to 1. -/
+def special_linear_group := { A : matrix n n R // A.det = 1 }
+
+end
+
+namespace special_linear_group
+
+variables {n : Type u} [fintype n] [decidable_eq n] {R : Type v} [comm_ring R]
+
+instance coe_matrix : has_coe (special_linear_group n R) (matrix n n R) :=
+⟨λ A, A.val⟩
+
+instance coe_fun : has_coe_to_fun (special_linear_group n R) :=
+{ F   := λ _, n → n → R,
+  coe := λ A, A.val }
+
+@[simp] lemma coe_fun_eq_val (A : special_linear_group n R) :
+  (A : n → n → R) = A.val :=
+rfl
+
+lemma ext_iff (A B : special_linear_group n R) : A = B ↔ (∀ i j, A i j = B i j) :=
+iff.trans subtype.ext ⟨(λ h i j, by simp [h]), matrix.ext⟩
+
+@[ext] lemma ext (A B : special_linear_group n R) : (∀ i j, A i j = B i j) → A = B :=
+(special_linear_group.ext_iff A B).mpr
+
+instance has_inv : has_inv (special_linear_group n R) :=
+⟨λ A, ⟨adjugate A, det_adjugate_eq_one A.2⟩⟩
+
+instance has_mul : has_mul (special_linear_group n R) :=
+⟨λ A B, ⟨A.1 ⬝ B.1, by erw [det_mul, A.2, B.2, one_mul]⟩⟩
+
+instance has_one : has_one (special_linear_group n R) :=
+⟨⟨1, det_one⟩⟩
+
+instance : inhabited (special_linear_group n R) := ⟨1⟩
+
+@[simp] lemma inv_val (A : special_linear_group n R) : A⁻¹.val = adjugate A.val := rfl
```

Comment:
Once you have set up the coercion to matrices, the idea would be to avoid `val` altogether, and write the lemmas using the coercion (for instance writing the left hand side of this lemma as `(A⁻¹ : matrix n n R)`.